### PR TITLE
BEP: Backstage Notifications System

### PR DESCRIPTION
**Pull request by Rugvip**
This is a [Backstage Enhancement Proposal (BEP)](https://github.com/backstage/backstage/tree/master/beps) for how to introduce a notifications system to Backstage. For questions on the BEP process, please refer to the [FAQ](https://github.com/backstage/backstage/tree/master/beps#faqs).

Note that merging this BEP does **not** mean there is a decision or agreement on the contents in the proposal. That point isn't until we move the BEP status to `implementable`. Please do **not** comment on the contents of the BEP in this PR, other than obvious editorial mistakes. Discussions and iteration of the proposal happens in a separate [discussion issue](https://github.com/backstage/backstage/issues/22213) and followup PRs.

The goal of this BEP is to align the work across #21728, #22099, and #21889. @drodil, @mareklibra, @ydayagi and the rest of Backstage core maintainers have all been consulted in the creation of this proposal.

### CODE REVIEW

File: beps/0001-notifications-system/README.md [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+For the signal plugin to reach a stable release we much reach the following:
+
+- A stable signal receiver filter format.
+- A stable signal channel API in the frontend.
+
+If any changes are required to the frontend framework to facilitate the implementation of notifications or signals, these will be released as experimental alpha features. They will stay in alpha until they are deemed stable enough, which must happen before a stable release of the notifications system.
+
+## Dependencies
+
+Since the signal plugin relies on the event broker for communication, it is a dependency for the notifications system as a whole. The event broker does not currently implement any transport for scaled deployments, which is a requirement for scaled deployments of the notification system.
```
Comment (drodil) [inline beps/0001-notifications-system/README.md:235]: Actually, in the current signals PR, the event broker is optional and if it's not passed to the `SignalService` it is assumed there's no scaled deployment in place and the signal will only be sent from that instance to all listeners.

File: beps/0001-notifications-system/README.md [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+
+## Alternatives
+
+### Signal Plugin Separation
+
+One primary consideration for the notifications system was whether the notification plugin should set up its own frontend push connection rather than relying on the signal plugin. The signal plugin separation is done for three primary reasons:
```
Comment (drodil) [inline beps/0001-notifications-system/README.md:241]: This is great stuff! I already have use cases for the signal plugin as it's own in my Q&A plugin

File: beps/0001-notifications-system/README.md [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+### Notifications Plugin
+
+The role of the notifications plugin is to manage the lifecycle of notifications. The backend plugin provides an API for other backends to send notifications, as well as an accompanying [backend service](https://backstage.io/docs/backend-system/architecture/services). It also provides a separate API for the frontend plugin to read notifications for an individual user and manage the read status of notifications.
+
+The notification backend stores notification using the [database service](https://backstage.io/docs/backend-system/core-services/index#database). In particular it needs to store the following information for each notification:
+
+- ID
+- Payload
+- Receivers
+- Read status by user
```
[the remark this answers is not in the document]
Comment (drodil) [reply beps/0001-notifications-system/README.md:90]: Also icons and images could be considered but they might be too much

File: beps/0001-notifications-system/README.md [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+
+The diagram below show an overview of the notifications system architecture and the interaction between the different components.
+
+![notifications system architecture diagram](./notifications-architecture.drawio.svg)
+
+The components that are presented with dashed lines are not directly part of this proposal, but included to be considered as future extensions.
+
+Let's walk through the process of sending of a notification to a user:
+
+1. A backend plugin that wants to send a notification to a user uses the `NotificationService`, which in turn makes a POST request to the notification backend. The request contains a filter to select the target users, as well as the notification payload.
```
Comment (drodil) [inline beps/0001-notifications-system/README.md:108]: What kind of filter is this? CatalogFilter or something similar? Implementation details IMO, but something that could be perhaps thought through

File: beps/0001-notifications-system/README.md [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+4. Each instance of the signal backend plugin subscribes to the `'signal'` topic and receives the event.
+5. Each signal backend instance has a set of push channels set up to online users. The incoming event is filtered based on the target user ID and the signaling is pushed through all connections with a matching user ID.
+6. The signal frontend plugin receives the signal and forwards it to all subscribers of the target channel. In this case, the notifications frontend plugin, who is subscribed to the `'notifications'` channel.
+7. The notifications frontend plugin receives the signal with the notification ID. It then makes a request to the notification backend to fetch the notification payload and then displays the notification to the user.
+
+   7b. Rather than rendering the notification directly, the notifications frontend plugin leaves that to an extension installed by the frontend part of the backend plugin that sent the notification. This allows for a more customized notification experience which can also contain custom actions and data display. This step is not in scope for this proposal.
+
+## Design Details
```
Comment (drodil) [inline beps/0001-notifications-system/README.md:121]: Maybe detail out also the database schema as it explains a lot about how the notifications should look like in all parts of the plugin?

File: beps/0001-notifications-system/README.md (hunk 4 above)
Comment (Rugvip) [reply beps/0001-notifications-system/README.md:108]: TBD in the details section

File: beps/0001-notifications-system/README.md (hunk 5 above)
Comment (Rugvip) [reply beps/0001-notifications-system/README.md:121]: Yep! The intention is for is all to fill out this section with more details in followup PRs

File: beps/0001-notifications-system/README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (Rugvip) [reply beps/0001-notifications-system/README.md:90]: Something to be discussed in the issue (see the PR body)

File: beps/0001-notifications-system/README.md (hunk 1 above)
Comment (Rugvip) [reply beps/0001-notifications-system/README.md:235]: That is also the default implementation of the event broker, so there's really no need to make it optional in the signal service imo.

File: beps/0001-notifications-system/README.md [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+
+The notification backend does not provide any new permissions, since creating notifications can only be done by other backend plugins, while reading notifications can only be done by the authenticated user. It is possible that we want to add a permissions for reading notifications, in particular for admin and impersonation use cases, but that is not part of this proposal or the initial implementation.
+
+The notification frontend plugin provides a UI for viewing notifications, which in the initial implementation can be as simple as needed. The only requirement is that a user is able to view recent notifications and distinguish between read and unread notifications. A notification as marked as read once it has been interacted with. The frontend plugin also subscribes to the notifications signal channel and alerts the user when a new notification is received.
```
Comment (mareklibra) [inline beps/0001-notifications-system/README.md:96]: The FE plugin should further provide FE API for other plugins to receive notifications.
Consuming FE plugins should be able to filter notifications they are interested in.

I will evolve this idea in the separate discussion issue.

Comment (mareklibra) [reply beps/0001-notifications-system/README.md:96]: https://github.com/backstage/backstage/issues/22213#issuecomment-1889106559

Comment (mareklibra) [reply beps/0001-notifications-system/README.md:96]: https://github.com/backstage/backstage/pull/22510

Comment (drodil) [reply beps/0001-notifications-system/README.md:96]: Some updates are also in #21889. The directions this takes are still too different to get to the implementation phase. Would it make sense to have another meetup to go into details and figure out what to do?

**Pull request merged**